### PR TITLE
feat: build and release binaries to GH releases

### PR DESCRIPTION
**Pull request by alecthomas**
Note that I can't really test this, but I successfully use something similar in an existing project.

### CODE REVIEW

File: .github/workflows/cli-release.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+name: Release CLI
+jobs:
+  build:
+    name: Build ${{ matrix.os }}-${{ matrix.architecture }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        architecture: [x86_64, aarch64]
+        include:
```
Comment (salman1993) [inline .github/workflows/cli-release.yml:18]: what are the additional configs here?

File: .github/workflows/cli-release.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          target: ${{ matrix.architecture }}-${{ matrix.suffix }}
+      - id: build
+        shell: bash
+        run: |
+          export TARGET=${{ matrix.architecture }}-${{ matrix.suffix }}
+          cargo build --release --target $TARGET
+          cd target/${TARGET}/release
+          tar -cJf goose-${TARGET}.tar.bz2 goose goosed
```
Comment (salman1993) [inline .github/workflows/cli-release.yml]: should this be lowercase `j` for bzip2? uppercase `J` uses xz compression. or maybe we rename to `.tar.xz`?

https://www.gnu.org/software/tar/manual/html_node/Option-Summary.html#Option-Summary

**Review by salman1993 — APPROVED** [review]
lgmt! left a comment to better understand the `include` config

**Comment by michaelneale** [comment]
@salman1993 @baxen how are binaries published at the moment? this would make a LOT of noise in releases if it was each push vs cutting a release? (but we need to have binaries published somewhere publicly)

**Comment by michaelneale** [comment]
(but could be the best approach if people don't mind many releases)

**Comment by alecthomas** [comment]
@michaelneale this only triggers on a tag matching the regex, and currently there are no binaries published AFAICT

File: .github/workflows/cli-release.yml (hunk 2 above)
Comment (alecthomas) [reply .github/workflows/cli-release.yml]: Fixed

File: .github/workflows/cli-release.yml (hunk 1 above)
Comment (alecthomas) [reply .github/workflows/cli-release.yml:18]: It's a mechanism for adding extra config for a matching matrix variable, documented [here](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#example-adding-configurations). In this case it's defining a Rust target suffix based on the OS we're  building on.

**Comment by michaelneale** [comment]
ok yes, I like this. hard to test so I guess we just... try it?

**Review by michaelneale — APPROVED** [review]

**Comment by michaelneale** [comment]
I haven't had much making this work - is there some source material I can look at for how people do it in rust?

**Pull request merged**